### PR TITLE
Remove blas

### PR DESCRIPTION
**Pull request by IvanPleshkov**
Remove blas from QDrant and use SIMD from std library. Works only for x86 and x64 arcs (because arm SIMD presents only in nightly rust).

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/spaces/simple.rs [hunk 1]
```diff
@@ -74,23 +60,111 @@ impl Metric for CosineMetric {
...
+    vector.iter().map(|x| x / length).collect()
+}
+
+fn dot_similarity(v1: &[VectorElementType], v2: &[VectorElementType]) -> ScoreType {
+    v1.iter().zip(v2).map(|(a, b)| a * b).sum()
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+#[target_feature(enable = "avx2")]
+unsafe fn hsum256_ps_avx(x: __m256) -> f32 {
```
Comment (generall) [inline lib/segment/src/spaces/simple.rs:105]: That looks pretty complicated! What is the source of this example?

File: lib/segment/src/spaces/simple.rs [hunk 2]
```diff
@@ -74,23 +60,111 @@ impl Metric for CosineMetric {
     }
 
     fn similarity(&self, v1: &[VectorElementType], v2: &[VectorElementType]) -> ScoreType {
-        v1.iter().zip(v2).map(|(a, b)| a * b).sum()
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            if is_x86_feature_detected!("avx2") {
```
Comment (generall) [inline lib/segment/src/spaces/simple.rs:65]: is it possible to also get some performance boost out of SSE* SIMD operations on older architectures?

File: lib/segment/src/spaces/simple.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/segment/src/spaces/simple.rs:105]: This function calculates the horizontal sum of all __m256 elements. I adopted an example from this conversation:
https://www.py4u.net/discuss/73145
Suggested implementation from the issue (https://github.com/zhao-lang/redis_hnsw/blob/master/src/hnsw/metrics.rs) has complicated __m256 sum too.

**Comment by generall** [comment]
Hi @IvanPleshkov, that's a great start for this issue :1st_place_medal:! I believe it would still require some extended benchmarking and testing on different machines before we could safely merge it. But here are some questions right away: 

- Does it require switching to nightly rust? (I saw https://doc.rust-lang.org/nightly/std/simd/index.html#)
- Is it possible to extent to other platform-specific implementations?
- How significantly performance is affected in case of available avx2?

File: lib/segment/src/spaces/simple.rs (hunk 2 above)
Comment (IvanPleshkov) [reply lib/segment/src/spaces/simple.rs:65]: It is possible to add AVX1 support. 10-years old processors have AVX2 instructions. Do we need support SIMD for older CPUs?

**Comment by IvanPleshkov** [comment]
> 
@generall it works on stable rust with AVX2. It is possible to support older x86/x64 CPUs. But ARM, MIPS and modern SSE with 512-bit support requires nightly rust
https://doc.rust-lang.org/core/arch/index.html#modules

I didn't measure performance yet. IMHO At now it is more important to decide on supported platforms and the way to solve the task. For instance, I can write a simple library in pure C with support on all platforms and bind it to QDrant through FFI.

**Comment by monatis** [comment]
IMHO support for Arm is more important than support for older X86/X64 processors, as Arm is gaining popularity in cloud computing.

It may be also used in embedded systems. I have plans to dockerize it for some demo apps on RPi with or without this PR already, so I'm always available on that side. It's a great starting point anyway.

**Comment by IvanPleshkov** [comment]
The point about cloud computing on ARM is good. But the restriction of this PR is nightly rust features. I hope that SIMD instructions for ARM and MIPS will come to stable rust earlier than ARM to most common servers.
@generall It is possible to implement ARM/MIPS/SSE4 over `cfg(nightly)`. This does not break stable rust build and achieves max perf on pure rust (but on nightly rust). We can remove cfg while features are coming to stable rust.

**Comment by generall** [comment]
> IMHO support for Arm is more important than support for older X86/X64 processors, as Arm is gaining popularity in cloud computing.

Unfortunately I have received a bunch of complains regarding support for the old CPU (apparently google cloud still allocates machines from 2010 sometimes :grimacing: ), but no complains regarding ARM support in a cloud.

For me priorities are the following:

1. It should be possible to run on any x86/x64 CPU without recompilation (from the same docker image)
2. It should be possible to build for ARM (there are a lot of developers using new M1 chip from Apple)
3. It should give max performance on top-tier hardware - i.e. able to use the most recent instructions
4. It should use available SIMD instructions on older CPUs

Current BLAS implementation solves p.3 only partially, cause it is not used in case of mmap storage. And this new approach solves this problem which I like pretty much.

**Comment by generall** [comment]
> It is possible to implement ARM/MIPS/SSE4 over cfg(nightly). This does not break stable rust build and achieves max perf on pure rust (but on nightly rust). We can remove cfg while features are coming to stable rust.

I am OK with using `nightly` in this specific case. Since it allows us to only enable this specific feature and avoid spreading unstable things around :smiley:

**Comment by IvanPleshkov** [comment]
@generall @monatis Added support or SSE (for old CPU from Pentium III). Current status of branch:
1) If x86 or x64 CPU supports avx2 instructions, avx2 256bit is used (actual from around 2011).
2) If x86 or x64 CPU supports SSE instructions, sse 128bit is used (actual for most of x86/x64 CPUs).
3) If CPU doesn't support SSE, AVX, or if it is not x86/x64 - trivial implementation without SIMD is used.

No specific compilations flags are needed, avx2/sse are checking in runtime.

Future work:
1) Add support of AVX-512 (for modern server CPUs). Restriction: nightly rust.
2) Add support of ARM NEON. Restriction: nightly rust.
3) Maybe another archs with SIMD like mips or powerpc is needed.
4) Test on a lot of various hardware and OS.

As the next step, I can add ARM NEON support for nightly rust build. I'm working from mac m1 and have the opportunity to test arm build.

**Comment by IvanPleshkov** [comment]
Important stuff. BLAS isn't used for euclidian metrics. In this branch, I added SIMD support for euclidian metric and for cosine preprocess

**Comment by generall** [comment]
https://github.com/calebzulawski/multiversion

**Review by generall — APPROVED** [review]

**Comment by generall** [comment]
:tada:

**Pull request merged**